src/System.IO.Pipes/tests/NamedPipeTests/NamedPipeTest.Specific.cs

```diff
@@ -0,0 +1,318 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System.Runtime.InteropServices;
+using System.Threading;
+using System.Threading.Tasks;
+using Xunit;
+
+namespace System.IO.Pipes.Tests
+{
+    /// <summary>
+    /// The Specific NamedPipe tests cover edge cases or otherwise narrow cases that
+    /// show up within particular server/client directional combinations.
+    /// </summary>
+    public class NamedPipeTest_Specific : NamedPipeTestBase
+    {
+        [Fact]
+        public void InvalidConnectTimeout_Throws_ArgumentOutOfRangeException()
+        {
+            using (NamedPipeClientStream client = new NamedPipeClientStream("client1"))
+            {
+                Assert.Throws<ArgumentOutOfRangeException>("timeout", () => client.Connect(-111));
+                Assert.Throws<ArgumentOutOfRangeException>("timeout", () => { client.ConnectAsync(-111); });
+            }
+        }
+
+        [Fact]
+        [ActiveIssue(812, PlatformID.AnyUnix)] // Unix implementation currently ignores timeout and cancellation token once the operation has been initiated]
+        public async Task ConnectToNonExistentServer_Throws_TimeoutException()
+        {
+            using (NamedPipeClientStream client = new NamedPipeClientStream(".", "notthere"))
+            {
+                var ctx = new CancellationTokenSource();
+                Assert.Throws<TimeoutException>(() => client.Connect(60));  // 60 to be over internal 50 interval
+                await Assert.ThrowsAsync<TimeoutException>(() => client.ConnectAsync(50));
+                await Assert.ThrowsAsync<TimeoutException>(() => client.ConnectAsync(60, ctx.Token)); // testing Token overload; ctx is not cancelled in this test
+            }
+        }
+
+        [Fact]
+        public async Task CancelConnectToNonExistentServer_Throws_TaskCanceledException()
+        {
+            using (NamedPipeClientStream client = new NamedPipeClientStream(".", "notthere"))
+            {
+                var ctx = new CancellationTokenSource();
+                if (RuntimeInformation.IsOSPlatform(OSPlatform.Windows)) // [ActiveIssue(812, PlatformID.AnyUnix)] - Unix implementation currently ignores timeout and cancellation token once the operation has been initiated
+                {
+                    Task clientConnectToken = client.ConnectAsync(ctx.Token);
+                    ctx.Cancel();
+                    await Assert.ThrowsAsync<TaskCanceledException>(() => clientConnectToken);
+                }
+                ctx.Cancel();
+                await Assert.ThrowsAsync<TaskCanceledException>(() => client.ConnectAsync(ctx.Token));
```

Comment:
This sporadically fails for me now that it's been changed away from ThrowsAnyAsync:
```
System.IO.Pipes.Tests.NamedPipeTest_Specific.CancelConnectToNonExistentServer_Throws_TaskCanceledException [FAIL]
        Assert.Throws() Failure
        Expected: typeof(System.Threading.Tasks.TaskCanceledException)
        Actual:   typeof(System.OperationCanceledException): The operation was canceled.
        Stack Trace:
              at System.Threading.CancellationToken.ThrowOperationCanceledException()
              at System.Threading.CancellationToken.ThrowIfCancellationRequested()
           \corefx\src\System.IO.Pipes\src\System\IO\Pipes\NamedPipeClientStream.cs(152,0): at System.IO.Pipes.NamedPipeClientStream
  .ConnectInternal(Int32 timeout, CancellationToken cancellationToken, Int32 startTime)
           \corefx\src\System.IO.Pipes\src\System\IO\Pipes\NamedPipeClientStream.cs(210,0): at System.IO.Pipes.NamedPipeClientStream
  .<>c__DisplayClass20_0.<ConnectAsync>b__0()
              at System.Threading.Tasks.Task.InnerInvoke()
              at System.Threading.Tasks.Task.Execute()
           --- End of stack trace from previous location where exception was thrown ---
              at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
              at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
              at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
              at System.Runtime.CompilerServices.TaskAwaiter.ValidateEnd(Task task)
              at System.Runtime.CompilerServices.TaskAwaiter.GetResult()
```